// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

syntax="proto2";
package brpc.policy;

option cc_generic_services = true;

message QuitRequest {
}

message QuitResponse {
}

message InitDBRequest {
    required string db_name = 1;
}

message InitDBResponse {
    // Set error_code to 0 if anything is ok, and db_name in request
    // will be set to MysqlConnContext.current_db by brpc.
    required int32 error_code = 1;
    optional string sql_state_marker = 2;
    optional string sql_state = 3;
    optional string error_message = 4;
}

message RefreshRequest {
    required bool grant = 1;
    required bool log = 2;
    required bool tables = 3;
    required bool hosts = 4;
    required bool status = 5;
    required bool threads = 6;
    required bool slave = 7;
    required bool master = 8;
}

message RefreshResponse {
}

message StatisticsRequest {
}

message StatisticsResponse {
    required string stats = 1;
}

message ProcessInfoRequest {
}

message ProcessInfoResponse {
}

message ProcessKillRequest {
    required uint32 connection_id = 1;
}

message ProcessKillResponse {
}

message DebugRequest {
}

message DebugResponse {
}

message PingRequest {
}

message PingResponse {
}

message QueryRequest {
    required bytes sql = 1;
}

message ResultMeta {
    optional string catalog = 1;
    optional string schema = 2;
    optional string table = 3;
    optional string org_table = 4;
    optional string name = 5;
    optional string org_name = 6;
    optional uint32 charset = 7;
    optional uint32 length = 8;
    optional uint32 type = 9;
    optional uint32 flags = 10;
    optional uint32 decimals = 11;
}

message RowField {
    optional bool is_null = 1;
    optional bytes field_value = 2;
}
message ResultRow {
    repeated RowField row_field = 1;
}

message QueryResponse {
    repeated ResultMeta meta = 1;
    repeated ResultRow row = 2;
    optional bool more_result = 3 [default = false];
}

message UnknownMethodRequest {
    required int32 command_id = 1;
}

message UnknownMethodResponse {
}

service MysqlService {
    rpc Quit(QuitRequest) returns (QuitResponse);
    rpc InitDB(InitDBRequest) returns (InitDBResponse);
    rpc Refresh(RefreshRequest) returns (RefreshResponse);
    rpc Statistics(StatisticsRequest) returns (StatisticsResponse);
    rpc ProcessInfo(ProcessInfoRequest) returns (ProcessInfoResponse);
    rpc ProcessKill(ProcessKillRequest) returns (ProcessKillResponse);
    rpc Debug(DebugRequest) returns (DebugResponse);
    rpc Ping(PingRequest) returns (PingResponse);
    rpc Query(QueryRequest) returns (QueryResponse);
    rpc UnknownMethod(UnknownMethodRequest) returns (UnknownMethodResponse);
}
